core/services/pipeline/graph.go

```diff
@@ -182,19 +95,94 @@ func (n *TaskDAGNode) Attributes() []encoding.Attribute {
 	return r
 }
 
-func (n *TaskDAGNode) outputs() []*TaskDAGNode {
-	var nodes []*TaskDAGNode
-	ns := n.g.From(n.ID())
-	for ns.Next() {
-		nodes = append(nodes, ns.Node().(*TaskDAGNode))
+//
+
+type Pipeline struct {
+	Tasks  map[int64]Task
+	tree   *Tree
+	Source string
+
+	// roots
+	// terminals / returns / outputs
+}
+
+func (p *Pipeline) UnmarshalText(bs []byte) (err error) {
+	p, err = Parse(bs)
+	return err
+}
+func (p *Pipeline) MinTimeout() (time.Duration, bool, error) {
+	var minTimeout time.Duration = 1<<63 - 1
+	var aTimeoutSet bool
+	for _, t := range p.Tasks {
+		if timeout, set := t.TaskTimeout(); set && timeout < minTimeout {
+			minTimeout = timeout
+			aTimeoutSet = true
+		}
 	}
-	return nodes
+	return minTimeout, aTimeoutSet, nil
 }
 
-func unwrapGraphNodes(nodes graph.Nodes) []*TaskDAGNode {
-	var out []*TaskDAGNode
-	for nodes.Next() {
-		out = append(out, nodes.Node().(*TaskDAGNode))
+func (p *Pipeline) TopoSort() []Task {
+	result := make([]Task, len(p.Tasks))
+
+	for id, task := range p.Tasks {
+		result[id] = task
 	}
-	return out
+	return result
+}
+
+func Parse(bs []byte) (*Pipeline, error) {
+	g := NewTree()
+	err := g.UnmarshalText(bs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Pipeline{
+		tree:   g,
+		Tasks:  make(map[int64]Task, g.Nodes().Len()),
+		Source: string(bs),
+	}
+
+	// toposort all the nodes: dependencies ordered before outputs. This also does cycle checking for us.
+	nodes, err := topo.SortStabilized(g, nil)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to topologically sort the graph, cycle detected")
+	}
+
+	for id, node := range nodes {
+		// use the new ordering as the id so that we can easily reproduce the original toposort
+		id := int64(id)
+
+		node, is := node.(*TreeNode)
+		if !is {
+			panic("unreachable")
+		}
+
+		if node.dotID == InputTaskKey {
+			return nil, errors.Errorf("'%v' is a reserved keyword that cannot be used as a task's name", InputTaskKey)
+		}
+
+		task, err := UnmarshalTaskFromMap(TaskType(node.attrs["type"]), node.attrs, id, node.dotID, nil, nil, nil)
```

Comment:
I've been thinking about dropping the last three params since they're always `nil`, we only set them right before execution. Thoughts?

Reply:
If possible to drop them, then do it.